README.md
# nrepl-adapter

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-05-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-05-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/nrepl-adapter/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/nrepl-adapter/compare/0.1.0...0.1.1


test/nrepl_adapter/core_test.clj
(ns nrepl-adapter.core-test
  (:require [clojure.test :refer :all]
            [nrepl-adapter.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/lib/dispatch.clj
;; event dispatching mechanism
;;
;; taken from ClojureScript One
;; https://github.com/brentonashworth/one
;; provided by Brenton Ashworth and Relevance, Inc
;; under the Eclipse Public License, the same as Clojure uses
;;

(ns ^{:doc "Event dispatching.

  Provides a way for code to react to events. Terminology:

  * event-id: Identifies a class of events. Can be any Clojure value.

  * event-data: Parameterizes a particular event. Can be any Clojure
    value.

  * reactor: A function that is invoked in response to an event
    occurring.

  * reaction: A relationship between a set of events and a reactor.

  * event-pred: A function which takes an event ID and returns true or
    false.

  Reactors are associated with events via `react-to`. When events are
  fired with an `event-id` and optional `event-data`, any reactors whose
  `event-pred` returns true for the `event-id` are invoked."}
  lib.dispatch)

(def ^{:doc "Stores the current reactions."}
  reactions (atom {}))

(defn react-to
  "Cause the specified reactor to be invoked whenever an event that
  satisfies `event-pred` is fired. reactor is a function that accepts
  two arguments: `event-id` and `event-data`.

  Returns the reaction.

  The reactor will continue to be invoked until one of two things
  happens:

   1. `delete-reaction` is called on this reaction.

   2. The reaction occurs `max-count` times. If `max-count` is not
      specified, the reaction will continue to be invoked until deleted.

  If `max-count` is specified, `delete-reaction` will be called
  automatically when the reaction has occurred the specified number of
  times."
  ([event-pred reactor]
     (react-to nil event-pred reactor))
  ([max-count event-pred reactor]
     (let [reaction {:max-count max-count
                     :event-pred event-pred
                     :reactor reactor}]
       (swap! reactions assoc reaction 0)
       reaction)))

(defn delete-reaction
  "Delete a reaction. After calling this function, the specified
  reaction will no longer be invoked."
  [reaction]
  (swap! reactions dissoc reaction))

(defn fire
  "Raise an event to any reactors whose event-pred returns true for
  `event-id`. The `event-id` and `event-data`, if specified, are passed to
  the reactor."
  ([event-id]
     (fire event-id nil))
  ([event-id event-data]
     (let [matching-reactions (filter (fn [[{event-pred :event-pred} run-count]]
                                        (event-pred event-id))
                                      @reactions)]
       (doseq [[reaction run-count] matching-reactions]
         (let [{:keys [max-count reactor]} reaction
               run-count (inc run-count)]
           (reactor event-id event-data)
           (if (and max-count
                    (<= max-count run-count))
             (delete-reaction reaction)
             (swap! reactions assoc reaction run-count)))))))

(comment

  (require '[cljs.repl :as repl])
  (require '[cljs.repl.rhino :as rhino])
  (repl/repl (rhino/repl-env))

  (do
    (let [recorded-reactions (atom [])
          reaction (react-to #{:do-something} #(swap! recorded-reactions conj [%1 %2]))]
      ;; Did we get a reaction back?
      (assert reaction)
      (fire :do-something)
      ;; Did the reactor catch the event?
      (assert (= [[:do-something nil]] @recorded-reactions))
      (fire :do-something)
      ;; Did the reactor catch the event a second time?
      (assert (= [[:do-something nil] [:do-something nil]] @recorded-reactions))
      (fire :something-else)
      ;; Did we ignore events we're not reacting to?
      (assert (= [[:do-something nil] [:do-something nil]] @recorded-reactions))
      (reset! recorded-reactions [])
      (fire :do-something 17)
      ;; Does event data arrive intact?
      (assert (= [[:do-something 17]] @recorded-reactions))
      (reset! recorded-reactions [])
      (delete-reaction reaction)
      (fire :do-something 17)
      ;; Does deleting a reaction cause us to stop reacting?
      (assert (= [] @recorded-reactions)))
    (let [recorded-reactions (atom #{})
          reaction-once (react-to 1 #{:do-something}
                                  #(swap! recorded-reactions conj [1 %1 %2]))
          reaction-twice (react-to 2 #{:do-something}
                                   #(swap! recorded-reactions conj [2 %1 %2]))]
      (fire :do-something 1)
      ;; Did both reactions react?
      (assert (= #{[1 :do-something 1] [2 :do-something 1]} @recorded-reactions))
      (fire :do-something 2)
      ;; Did only the second reaction react?
      (assert (= #{[1 :do-something 1] [2 :do-something 1] [2 :do-something 2]}
                 @recorded-reactions))
      (fire :do-something 3)
      ;; Did nothing change?
      (assert (= #{[1 :do-something 1] [2 :do-something 1] [2 :do-something 2]}
                 @recorded-reactions)))
    true
    )
  )


src/lib/simple_tcp_connection/client.clj
; simple tcp connection library
;;
;; by Otto Linnemann
;; (C) 2017, GNU General Public Licence

(ns lib.simple-tcp-connection.client
  (:require [clojure.core.async
             :as a
             :refer [>! <! >!! <!! go chan buffer close! thread
                     alts! alts!! timeout]]
            [lib.dispatch :as dispatch])
  (:use [lib.simple-tcp-connection.send-receive]
        lib.simple-tcp-connection.utils)
  (:import [java.net Socket]))


(comment
  (def s (Socket. "localhost" 5044))
  (. s setSoTimeout 3000)

  (ssend s "hello world\n")
  (sreceive s)
  )


(defn- pr-debug
  "additional output for debugging purposes"
  [msg]
  (comment println msg))


(defn- create-tcp-client-connection
  "tries to create  a tcp client connection to the  specified server address and
  port. Instead  of throwing an  exception when the  client does not  accept the
  connection return nil."
  [addr port]
  (try
    (do
      (pr-debug (str "-------> invoke tcp-client with "  addr ", " port))
      (Socket. addr port))
    (catch Exception e
        (println "could not connect to server on address: " addr ", port: " port " error!\n")
      (Thread/sleep 100)
      nil)))

; (create-tcp-client-connection "localhost" 5044)


(defn- check-and-update-client-connection
  "checks the reference of the TCP client connection (to a server). If it
   is invalidated reconnect and fire a connection event."
  [conref addr port con-event]
  (swap!
   conref
   (fn [c] (or c
               (let [c (create-tcp-client-connection addr port)]
                 (when c (dispatch/fire con-event))
                 c)))))


(defn create-tcp-server-request-process
  "Creates a asynchronous  background thread which reads requests  from an input
  channel, sends  these to the specified  tcp server with specified  address and
  port and waits  at max the speficied  delay time for a response.  When a delay
  time  of zero  is  given the  requests  are  processed in  a  fire and  forget
  operation mode.

  The background process continously keeps trying to setup the TCP connection to
  the server to cover situation where the server crashed and is restarted.

  The  function  returns  a  hash  map   with  the  input  and  output  channels
  for  requests  and  responses  and  among  other  data  and  event  key  where
  external  listeners  might   registering  to  be  informed   e.g.  also  about
  unsolicited events.  The event  loop itself is  implemented with  the function
  start-tcp-server-response-event-loop which needs to be invoked separately."
  [addr port wait-for-response-ms]
  (let [in (chan)
        out (chan)
        conref (atom nil)
        run (atom true)
        resp-event (gensym :received-response)
        con-event (gensym :server-connected)
        discon-event (gensym :server-disconnected)]
    (go
      (loop []
        (let [request (<! in)
              con (check-and-update-client-connection conref addr port con-event)
              timeout-ch (timeout wait-for-response-ms)
              resp-reactor (dispatch/react-to 1 #{resp-event} (fn [evt data] (>!! timeout-ch data)))
              send-result (if con (ssend con request) -99)]
          (if (< send-result 0)
            (do (reset! conref nil)
                (>! out false))
            (do (if (= wait-for-response-ms 0)
                  (>! out true)
                  (>! out (if-let [v (<! timeout-ch)] v false)))))
          (dispatch/delete-reaction resp-reactor))
        (when @run
          (Thread/sleep 5000)
          (recur))))
    (hash-args addr port in out conref run resp-event con-event discon-event)))



(defn start-tcp-server-response-event-loop
  "starts the  background response  process event  loop for  a given  tcp server
  request process."
  [process]
  (go
    (let [{:keys [addr port conref run resp-event con-event discon-event]} process]
      (loop []
        (pr-debug ".... --- entering loop --- ....")
        (if-let [con (check-and-update-client-connection conref addr port con-event)]
          (if-let [resp (try (sreceive con)
                             (catch Exception e
                               (reset! conref nil)
                               (println "read descriptor for address:" addr " became invalid!\n")
                               (dispatch/fire discon-event)
                               nil))]
            (do
              (pr-debug (str resp))
              (dispatch/fire resp-event resp)
              (when @run (recur)))
            (when @run
              (Thread/sleep 1000)
              (recur)))
          (when @run
            (Thread/sleep 1000)
            (recur)))))))



(defn release-tcp-server-processes
  "stops all background threads of the given process"
  [process]
  (reset! (:run process) false)
  (when-let [con @(:conref process)]
    (.close con)))



(defn create-tcp-server-connection-handler
  "shortcut      to      invoke     'create-tcp-server-request-process'      and
  'start-tcp-server-response-event-loop'.  Function   initializes  all  required
  background  handlers and  channels in  order to  maintain communication  to an
  external TCP server which is listening on the specified address and port. When
  no timeout value is specified a default value of 500ms wait time for responses
  is  used.  Specifiy other  values  or  zero  in  order to  completely  disable
  response processing. Refer  to the aforementioned functions  for more detailed
  information."
  ([addr port]
   (create-tcp-server-connection-handler addr port 500))
  ([addr port wait-for-response-ms]
   (let [process (create-tcp-server-request-process addr port wait-for-response-ms)]
     (start-tcp-server-response-event-loop process)
     process)))


(defn send-tcp-server-request
  "sends  request e.g.  remote procedure  call  to given  processor and  returns
  either the answer from the server or false when no answer has been received or
  the connection to the server is broken. When the background request process is
  in fire and forget mode the function  returns 'true' when the request could be
  successfully send, otherwise false."
  [process request]
  (let [in (:in process)
        out (:out process)]
    (>!! in request)
    (<!! out)))


(defn react-to-tcp-server-response
  "registers a callback handler which is invoked for every response from the TCP
  server which  is described  with the given  hash-table 'process'.  The handler
  needs to  process two  arguments, the  first one being  the event  itself, the
  second one provides the response data from the server. The function returns an
  event reactor  handle. The callback  can be  disable by invoking  the function
  'dispatch/delete-reaction'."
  [process cb]
  (dispatch/react-to #{(:resp-event process)} cb))


(defn react-to-tcp-server-connection-event
  "registers a callback handler which is  invoked when the connection to the TCP
  server is established."
  [process cb]
  (dispatch/react-to #{(:con-event process)} cb))


(defn react-to-tcp-server-disconnection-event
  "registers  a callback  handler  which is  invoked when  the  client has  been
  disconneted from TCP server."
  [process cb]
  (dispatch/react-to #{(:discon-event process)} cb))


(comment
  ;; usage illustration

  (def tst-process (create-tcp-server-connection-handler "127.0.0.1" 7777 5000))
  (def tst-process (create-tcp-server-connection-handler "127.0.0.1" 7777 0))
  (def resp (send-tcp-server-request tst-process "sample command\n"))

  (release-tcp-server-processes tst-process)

  (def response-reactor
    (react-to-tcp-server-response
     tst-process
     (fn [evt data] (println "---> received evt: " data))))

  (def connection-reactor
    (react-to-tcp-server-connection-event
     tst-process
     (fn [evt data] (println "---> connected to server"))))

  (def disconnection-reactor
    (react-to-tcp-server-disconnection-event
     tst-process
     (fn [evt data] (println "---> disconnected from server"))))

  (dispatch/delete-reaction response-reactor)
  (dispatch/delete-reaction connection-reactor)
  (dispatch/delete-reaction disconnection-reactor)

  (release-tcp-server-processes tst-process)

  )


src/nrepl_adapter/core.clj
(ns nrepl-adapter.core
  (:require [clojure.tools.nrepl.transport :as t]
            [clojure.tools.cli :refer [parse-opts]]
            [clojure.java.io :as io]
            [clojure.string :as str])
  (:use [clojure.tools.nrepl.server :only [start-server stop-server]]
        [clojure.tools.nrepl.misc :only [response-for uuid]]
        [clojure.tools.nrepl.middleware.pr-values :only [pr-values]]
        [lib.simple-tcp-connection.client])
  (:import java.util.Properties
           [java.net InetAddress])
  (:gen-class))


(def ^{:private true} sessions (atom {}))


(defn get-version
  "reads version data from the pom.properties META-INF file
   refer to:
   https://groups.google.com/d/msg/leiningen/7G24ifiYvOA/h6xmjeWaO3gJ"
  [dep]
  (let [path (str "META-INF/maven/" (or (namespace dep) (name dep))
                  "/" (name dep) "/pom.properties")
        props (io/resource path)]
    (when props
      (with-open [stream (io/input-stream props)]
        (let [props (doto (Properties.) (.load stream))]
          (.getProperty props "version"))))))


(defn- create-session
  "Returns a new atom containing a map of bindings as per
   `clojure.core/get-thread-bindings`.  Values for *out*, *err*, and *in*
   are obtained using `session-in` and `session-out`, *ns* defaults to 'user,
   and other bindings as optionally provided in `baseline-bindings` are
   merged in."
  [& {:keys [transport baseline-bindings client-addr client-port]
      :or {baseline-bindings {}}}]
  (let [client-handle
        (when (and client-addr client-port) (create-tcp-server-connection-handler client-addr client-port 5000))]
    (if client-handle
      (atom (merge baseline-bindings {:client-handle client-handle}))
      (atom baseline-bindings))))


(defn- register-session
  [msg]
  (let [{:keys [transport session]} msg
        id (uuid)]
    (let [session (create-session :transport transport :baseline-bindings @session)]
      (swap! sessions assoc id session)
      (t/send transport (response-for msg :status :done :new-session id)))))


(defn- close-session
  [msg]
  (let [{:keys [transport session id]} msg]
    (release-tcp-server-processes (:client-handle @session))
    (swap! sessions dissoc id)
    (t/send transport (response-for msg :status #{:done :session-closed}))))


(defn- split-result-from-prompt
  "returns vector with result and last line which is the prompt"
  [s]
  (let [lines (str/split s #"\n")]
    [(apply str (interpose "\n" (butlast lines)))
     (-> (last lines)
         (str/replace #"[=>].*$" "")
         (str/replace #"[<]" "")
         (str/trim))]))


(defn- strip-cr-and-comments
  "strips out all carriage return linefeed characters and
   remove comments"
  [s]
  (apply str
         (map
          #(str/replace % #"[;].*$" "")
          (str/split s #"[\n\r]"))))


(defn- strip-namespace-tags
  "strip out pixie's namespace tag"
  [s]
  (-> s
      (str/replace #"<Namespace" "")
      (str/replace #"[>]" "")
      (str/trim)))


(defn- bind-inferior-repl-handler-connection
  [client-addr client-port]
  (fn [{:keys [op ns session code file line column id transport] :as msg}]
    (let [the-session (if session
                        (@sessions session)
                        (create-session :transport transport
                                        :client-addr client-addr :client-port client-port))]
      (if-not the-session
        (t/send transport (response-for msg :status #{:error :unknown-session}))
        (let [msg (assoc msg :session the-session)]
          (case op
            "clone" (register-session msg)
            "close" (close-session msg)
            "describe"
            (do (t/send transport (response-for msg :status :done :ops {:eval {}})))
            "eval"
            (let [client-handle (:client-handle @the-session)
                  resp (if (re-find #"cider.nrepl|clojure.core" code)
                         "error"
                         (send-tcp-server-request client-handle (strip-cr-and-comments code)))
                  [result ns] (if resp
                                (split-result-from-prompt resp)
                                ["could not connect to telnet REPL error!" ""])]
              (t/send transport
                      (response-for msg :value
                                    (let [code (format "(pr-str %s)" code)]
                                      (if (re-find #"\*ns\*" code)
                                        (pr-str (strip-namespace-tags result))
                                        result)
                                      )))
              (t/send transport
                      (if resp
                        (response-for msg :status :done :ns ns)
                        (response-for msg :status #{:error :unknown-session :done}))))))))))


(defn- start-inf-repl-adapter
  [server-port client-addr client-port]
  (def server (start-server :port server-port :handler
                            (bind-inferior-repl-handler-connection client-addr client-port))))

; (start-inf-repl-adapter 9000 "localhost" 37147)
; (stop-server server)


(defn str2port-num
  "simple string to integer converter"
  [s]
  (try
    (let [port (.intValue (Integer. s))]
      (if (and (> port 0) (< port 65536)) port nil))
    (catch NumberFormatException e nil)))


; command line interface (leiningen)
(def cli-options
  [["-s" "--server-port PORT" "port where the nrepl is listening to"
    :default 37148
    :validate [#(str2port-num %) "TCP port must be in valid range [1..65535]"]]
   ["-c" "--client-port PORT" "port where the the adapter is connecting to (inferior repl)"
    :default 37147
    :validate [#(str2port-num %) "TCP port must be in valid range [1..65535]"]]
   ["-a" "--client-addr ADDR" "TCP address where the client is connecting to."
    :default "localhost"
    :parse-fn #(when (InetAddress/getByName %) %)]
   ["-h" "--help" "this help string"]])


(defn- cli
  [& args]
  (let [opts (parse-opts args cli-options)
        options (:options opts)
        arguments (:arguments opts)
        summary (:summary opts)
        errors (:errors opts)
        invalid-opts (not-empty errors)
        title-str (str
                   "nrepl-adapter: nREPL to inferior Shell REPL Adapter\n"
                   (format "      Version: %s, refer to https://github.com/linneman/nrepl-adapter for more information\n" (get-version 'nrepl-adapter))
                   "      (C) 2017, GNU General Public Licence by Otto Linnemann\n")
        {:keys [server-port client-port client-addr]} options
        [server-port client-port] (map str2port-num [server-port client-port])]
    (println title-str)
    (if (or (:help options) invalid-opts)
      (do
        (if invalid-opts
          (println errors))
        (println "  Invocation:\n")
        (println summary)
        -1)
      (do
        (println (format "Start listening on port %s, forwarding s-expr to addr: %s, port: %s ..."
                         server-port client-addr client-port))
        (start-inf-repl-adapter server-port client-addr client-port)))))


(comment
  (cli "-s" "5000")
  (cli "-s" "5000" "--client-port" "3000" "--client-addr" "maxwell")
  (cli "-s" "5000" "--client-port" "3000" "-a" "maxwell")
  (cli "-a" "xxx")
  (cli "--help")
  (cli "-h")
  (cli)
  )

(defn -main [& args]
  (try
    (apply cli args)
    (catch Throwable t
      (let [msg (format "%s" (.getMessage t))
            cause (.getCause t)
            st (.getStackTrace t)]
        (println msg)
        (when cause (println "cause: " cause))
        (dorun (map #(println %) st))))))
